File: Cheapflight.java
import java.util.*;

class AlgoJet {
    private Map<String, Map<String, Integer>> flights = new HashMap<>();

    public void addFlight(String source, String destination, int price) {
        flights.computeIfAbsent(source, k -> new HashMap<>()).put(destination, price);
        flights.computeIfAbsent(destination, k -> new HashMap<>()).put(source, price);
    }

    public int findCheapestFlight(String source, String destination) {
        if (!flights.containsKey(source) || !flights.containsKey(destination)) {
            return -1; // Cities are not in the database
        }

        Map<String, Integer> cheapestPrices = new HashMap<>();
        for (String city : flights.keySet()) {
            cheapestPrices.put(city, Integer.MAX_VALUE);
        }

        cheapestPrices.put(source, 0);
        Set<String> visited = new HashSet<>();

        while (true) {
            String currentCity = null;
            int currentPrice = Integer.MAX_VALUE;

            // Find the city with the cheapest known price
            for (String city : flights.keySet()) {
                if (!visited.contains(city) && cheapestPrices.get(city) < currentPrice) {
                    currentCity = city;
                    currentPrice = cheapestPrices.get(city);
                }
            }

            if (currentCity == null) {
                break; // No more cities to visit
            }

            visited.add(currentCity);

            for (Map.Entry<String, Integer> neighbor : flights.get(currentCity).entrySet()) {
                String neighborCity = neighbor.getKey();
                int neighborPrice = neighbor.getValue();
                int totalPrice = currentPrice + neighborPrice;

                if (totalPrice < cheapestPrices.get(neighborCity)) {
                    cheapestPrices.put(neighborCity, totalPrice);
                }
            }
        }

        return cheapestPrices.get(destination);
    }

    public static void main(String[] args) {
        AlgoJet algoJet = new AlgoJet();

        // Add flights to the system
        algoJet.addFlight("A", "B", 100);
        algoJet.addFlight("A", "C", 150);
        algoJet.addFlight("B", "C", 40);
        algoJet.addFlight("B", "D", 200);
        algoJet.addFlight("C", "D", 100);
        algoJet.addFlight("C", "E", 120);
        algoJet.addFlight("D", "E", 80);

        // Find and print the cheapest flight from A to E
        int cheapestPrice = algoJet.findCheapestFlight("A", "D");
        System.out.println("Cheapest flight from A to E: " + cheapestPrice);
    }
}

File: cheapflight.py
import sys

class Flight:
    def __init__(self, source, destination, price):
        self.source = source
        self.destination = destination
        self.price = price

class FlightGraph:
    def __init__(self):
        self.flights = []

    def initializeFlightGraph(self, flights):
        self.flights = flights

    def getCheapestFlight(self, source, destination):
        cities = set()
        for flight in self.flights:
            cities.add(flight.source)
            cities.add(flight.destination)

        distances = {city: sys.maxsize for city in cities}
        distances[source] = 0

        for _ in range(len(cities)):
            for flight in self.flights:
                if distances[flight.source] + flight.price < distances[flight.destination]:
                    distances[flight.destination] = distances[flight.source] + flight.price

        if distances[destination] == sys.maxsize:
            return -1
        else:
            return distances[destination]

# Example usage:
flights = [
    Flight("A", "B", 100),
    Flight("A", "C", 150),
    Flight("B", "C", 40),
    Flight("B", "D", 200),
    Flight("C", "D", 100),
    Flight("C", "E", 120),
    Flight("D", "E", 80)
]

flight_graph = FlightGraph()
flight_graph.initializeFlightGraph(flights)

source = input('> ')
destination = input('> ')
cheapest_price = flight_graph.getCheapestFlight(source, destination)

print(f"The cheapest flight from {source} to {destination} costs {cheapest_price}")
